infra/blockchain/evmErrors.go

```diff
@@ -0,0 +1,193 @@
+package blockchain
+
+import (
+	"fmt"
+	"math/big"
+	"strings"
+)
+
+type ErrNotEnoughCashes map[string]ErrNotEnoughCash
+
+func (e ErrNotEnoughCashes) Error() string {
+	msgs := []string{}
+	for user, err := range e {
+		msgs = append(msgs, fmt.Sprintf("%v %v", user, err.Error()))
+	}
+	return strings.Join(msgs, "\n")
+}
+
+type ErrNotEnoughCash struct {
+	Need *big.Int
+	Got  *big.Int
+}
+
+func (e ErrNotEnoughCash) Error() string {
+	return fmt.Sprintf("out of balance, need %v, got %v", e.Need, e.Got)
+}
+
+var (
+	TX_PENDING_REASON_NOT_ENOUGH_CASH_BUT_NOT = "notEnoughCash but balance enough"
+	TX_PENDING_REASON_NOT_ENOUGH_CASH         = "notEnoughCash"
+	TX_PENDING_REASON_FUTURE_NONCE            = "futureNonce"
+	TX_PENDING_REASON_OLD_EPOCH_HEIGHT        = "oldEpochHeight"
+	TX_PENDING_REASON_OUT_DATED_STATUS        = "outdatedStatus"
+)
+
+type TxRpcError int
+
+const (
+	TX_ERR_RPC_OUT_OF_BALANCE TxRpcError = iota
+	TX_ERR_RPC_TXPOOL_FULL
+	TX_ERR_RPC_INSERT_VALIDATION_FAIL
+	TX_ERR_RPC_INSERT_FAIL_QUOTA_UNENOUGH
+	TX_ERR_RPC_SAME_NONCE_ALREADY_INSERTED
+	TX_ERR_RPC_STALE_NONCE
+	TX_ERR_RPC_EXCEED_MAX_GAS
+	TX_ERR_RPC_POOL_INCONSISTENT
+	TX_ERR_RPC_FAIL_READ_ACCOUNT_CACHE
+	TX_ERR_RPC_NONCE_DISTANT
+	TX_ERR_RPC_FAILED_IMPORT_TO_DEFFER_POOL
+	TX_ERR_RPC_TX_ALREADY_EXIST
+
+	TX_ERR_RPC_TOO_MANY_REQUEST
+	TX_ERR_RPC_CONFURA_QPS_EXCEED
+	TX_ERR_RPC_CONFURA_DAILYQ_EXCEED
+	TX_ERR_RPC_MISSING_BEST_HASH
+	TX_ERR_RPC_EPOCH_HEIGHT_OUT_BOUND
+	// 包括 vm revert, not enough cash
+	TX_ERR_ESTIMATE_OTHERS
+	TX_ERR_RPC_ESTIMATE_NOT_ENOUGH_CASH
+	TX_ERR_RPC_ESTIMATE_VM_ERR
+	TX_ERR_DROPPED
+	TX_ERR_RPC_OTHER
+)
+
+func (e *TxRpcError) NeedResendWhenSentErr() (needSend, needUpperGas bool) {
```

Comment:
The function parameter e is a pointer to TxRpcError, but it is not modified inside the function. It is better to use a value receiver instead of a pointer receiver for such functions, as it avoids unnecessary indirection and nil checks. See https://golang.org/doc/faq#methods_on_values_or_pointers for more details.

Reply:
The function name NeedResendWhenSentErr is not clear and does not follow the convention of starting with a verb. Consider renaming it to something like ShouldResendOrIncreaseGas or CheckTxErrorStatus.